src/fivetran-destination/src/utils.rs

```diff
@@ -98,3 +103,262 @@ pub fn to_fivetran_type(ty: Type) -> Result<(DataType, Option<DecimalParams>), O
         }
     }
 }
+
+/// Maps the column ordering of a CSV file, to the ordering of the provided table.
+///
+/// The column ordering of a CSV provided by Fivetran is not guaranteed to match the column
+/// ordering of the destination table, so based on column name we re-map the CSV.
+#[derive(Debug)]
+pub struct AsyncCsvReaderTableAdapter<R> {
+    /// Async CSV reader.
+    csv_reader: csv_async::AsyncReader<R>,
+    /// The `i`th column in a table, exists at the `tbl_to_csv[i]` position in a CSV record.
+    tbl_to_csv: Vec<usize>,
+}
+
+impl<R> AsyncCsvReaderTableAdapter<R>
+where
+    R: AsyncRead + Unpin + Send + 'static,
+{
+    /// Creates a new [`AsyncCsvReaderTableAdapter`] from an [`AsyncRead`]-er and a [`Table`].
+    pub async fn new(reader: R, table: &Table) -> Result<Self, OpError> {
+        let mut csv_reader = csv_async::AsyncReaderBuilder::new().create_reader(reader);
+        let csv_headers = csv_reader.headers().await?;
+        let csv_columns: BTreeMap<_, _> = csv_headers
+            .iter()
+            .enumerate()
+            .map(|(idx, name)| (name.trim().to_string(), idx))
+            .collect();
+
+        // TODO(parkmycar): When we support the `mz_extras` column we'll need to handle this case.
+        if table.columns.len() != csv_columns.len() {
+            let msg = format!(
+                "number of columns do not match: table {}, csv {}",
+                table.columns.len(),
+                csv_columns.len()
+            );
+            return Err(OpErrorKind::CsvMapping {
+                headers: csv_headers.clone(),
+                table: table.clone(),
+                msg,
+            }
+            .into());
+        }
+
+        // Create our column mapping.
+        let tbl_to_csv = table
+            .columns
+            .iter()
+            .map(|col| {
+                csv_columns.get(&col.name).copied().ok_or_else(|| {
```

Comment:
What about `.remove` here instead, then verify at the end the btreemap is empty? That'd guarantee that all columns are used exactly once.

Reply:
Good call! Added